language_service/src/definition.rs

```diff
@@ -40,43 +42,185 @@ pub(crate) fn get_definition(
+                                }
+                            }
+                            _ => panic!("UDT has invalid resolution."),
                         }
                     }
-                    Res::Local(_node) => None,
-                };
-                if let Some(def) = item {
-                    if let ItemKind::Callable(decl) = &def.kind {
-                        self.definition = Some((
-                            self.source_map
-                                .find_by_offset(decl.name.span.lo)
-                                .expect("source should exist for offset")
-                                .name
-                                .to_string(),
-                            decl.name.span.lo,
-                        ));
+                }
+                _ => walk_expr(self, expr),
+            }
+        }
+    }
+
+    // Handles local variable, UDT, and callable references
+    fn visit_path(&mut self, path: &'_ ast::Path) {
+        if span_contains(path.span, self.offset) {
+            let res = self.compilation.unit.ast.names.get(path.id);
+            if let Some(res) = res {
+                match &res {
+                    resolve::Res::Item(item_id) => {
+                        if item_id.package.is_none() {
+                            if let Some(item) = find_item(self.compilation, item_id) {
+                                let lo = match &item.kind {
+                                    hir::ItemKind::Callable(decl) => decl.name.span.lo,
+                                    hir::ItemKind::Namespace(_, _) => {
+                                        panic!(
+                                            "Reference node should not refer to a namespace: {}",
+                                            path.id
+                                        )
+                                    }
+                                    hir::ItemKind::Ty(ident, _) => ident.span.lo,
+                                };
+                                self.set_definition_from_position(lo);
+                            };
+                        }
                     }
+                    resolve::Res::Local(node_id) => {
+                        if let Some(curr) = self.curr_callable {
+                            {
+                                let mut finder = AstFindNode {
+                                    node_id,
+                                    result: None,
+                                };
+                                finder.visit_callable_decl(curr);
+                                if let Some(lo) = finder.result {
+                                    self.set_definition_from_position(lo);
+                                }
+                            }
+                        }
+                    }
+                    _ => {}
                 }
             }
         }
-        walk_expr(self, expr);
+    }
+}
+
+struct AstFindNode<'a> {
```

Comment:
Nit: I think the name of this struct is deceptively generic - it only works on `Pat` node kinds. Maybe `PatFinder` or something?

Reply:
Renamed to `AstPatFinder`